core/eolearn/core/utils/parallelize.py

```diff
@@ -0,0 +1,208 @@
+    :return: A list of results in the same order as input parameters given by `executor_params`.
+    """
+    futures = [executor.submit(function, *function_params) for function_params in zip(*params)]
+    return join_futures(futures, **tqdm_kwargs)
+
+
+def join_futures(futures: List[Future], **tqdm_kwargs: Any) -> List[Any]:
+    """Resolves futures, monitors progress, and returns a list of results.
+
+    :param futures: A list of futures to be joined. Note that this list will be reduced into an empty list as a side
+        effect of this function. This way future objects will get cleared from memory already during the execution
+        which will free some extra memory. But this can be achieved only if future objects aren't kept in memory
+        outside `futures` list.
+    :param tqdm_kwargs: Keyword arguments that will be propagated to `tqdm` progress bar.
+    :return: A list of results in the order that corresponds with the order of the given input `futures`.
+    """
+    results: List[Optional[Any]] = [None] * len(futures)
+    for position, result in join_futures_iter(futures, **tqdm_kwargs):
+        results[position] = result
+
+    return cast(List[Any], results)
+
+
+def join_futures_iter(
+    futures: List[Future], update_interval: float = 0.5, **tqdm_kwargs: Any
+) -> Generator[Tuple[int, Any], None, None]:
+    """Resolves futures, monitors progress, and serves as an iterator over results.
+
+    :param futures: A list of futures to be joined. Note that this list will be reduced into an empty list as a side
+        effect of this function. This way future objects will get cleared from memory already during the execution
+        which will free some extra memory. But this can be achieved only if future objects aren't kept in memory
+        outside `futures` list.
+    :param update_interval: A number of seconds to wait between consecutive updates of a progress bar.
+    :param tqdm_kwargs: Keyword arguments that will be propagated to `tqdm` progress bar.
+    :return: A generator that will be returning pairs `(index, result)` where `index` will define the position of future
+        in the original list to which `result` belongs to.
+    """
+
+    def _wait_function(
+        remaining_futures: Iterable[Future], timeout: float
+    ) -> Tuple[Iterable[Future], Iterable[Future]]:
+        done, not_done = concurrent.futures.wait(remaining_futures, timeout=timeout, return_when=FIRST_COMPLETED)
+        return done, not_done
+
+    def _get_result(future: Future) -> Any:
+        return future.result()
+
+    return _base_join_futures_iter(_wait_function, _get_result, futures, update_interval, **tqdm_kwargs)
+
+
+def _base_join_futures_iter(
+    wait_function: Callable[[Iterable[_InputType], float], Tuple[Iterable[_InputType], Iterable[_InputType]]],
+    get_result_function: Callable[[_InputType], _OutputType],
+    futures: List[_InputType],
+    update_interval: float = 0.5,
```

Comment:
feels to me like this can be "hardcoded" in the closures you create in `join_futures_iter` (and same for ray).